weather_forecast_retrieval/hrrr_nomads.py

```diff
@@ -0,0 +1,193 @@
+
+
+class HRRRNOMADS():
+    """Helper class to aid in the interaction with downloading HRRR
+    data from NOMADS.
+    """
+
+    def __init__(self, output_dir, num_requests=2, overwrite=False, verbose=False) -> None:
+
+        logLevel = 'DEBUG' if verbose else 'INFO'
+        self._logger = utils.setup_local_logger(__name__, loglevel=logLevel)
+
+        self.output_dir = output_dir
+        self.num_requests = num_requests
+        self.overwrite = overwrite
+        self.verbose = verbose
+
+        self.http_retrieval = HttpRetrieval(overwrite=overwrite, external_logger=self._logger)
+        self.http_retrieval.output_dir = output_dir
+
+    def set_dates(self, start_date, end_date, forecast_hrs):
+
+        self.start_date = start_date
+        self.end_date = end_date
+        self.forecast_hrs = forecast_hrs
+
+    def date_range(self, start_date, end_date, forecast_hrs=None):
+        """Get files within a date range.
+
+        Args:
+            start_date (str): start date of period
+            end_date (str): end date of period
+            forecast_hrs (list, optional): Forecast hours to download.
+                Defaults to None or download all hours.
+
+        Returns:
+            list: List of response objects
+        """
+
+        self.set_dates(start_date, end_date, forecast_hrs)
+        return self.http_retrieval.fetch_by_date(start_date, end_date, forecast_hrs)
+
+    def latest(self, latest=3, forecast_hrs=None):
+        """Download files for the lastest hours.
+
+        Args:
+            latest (int, optional): Number of hours to download. Defaults to 3.
+            forecast_hrs (list, optional): Forecast hours to download.
+                Defaults to None or download all hours.
+
+        Returns:
+            list: List of response objects
+        """
+
+        end_date = datetime.utcnow().replace(microsecond=0)
+        start_date = end_date - timedelta(hours=latest)
+        self.set_dates(start_date, end_date, forecast_hrs)
+
+        return self.http_retrieval.fetch_by_date(start_date, end_date, forecast_hrs)
+
+    def preprocessing(self, bbox, output_path):
+        """Preprocess the downloaded files by cropping and extracting to the
+        output_path
+
+        Args:
+            bbox (list): Bounding box [lon W, lon E, lat S, lat N]
+            output_path (str): directory to put cropped files
+
+        Raises:
+            ValueError: forecast_hrs must be set explicitly
+        """
+
+        if self.forecast_hrs is None:
+            raise ValueError('Must specify forecast hours when preprocessing')
+
+        for fx_hr in self.forecast_hrs:
+
+            HRRRPreprocessor(
+                self.output_dir,
+                self.start_date,
+                self.end_date,
+                output_path,
+                bbox,
+                fx_hr,
+                verbose=self.verbose
+            ).run()
+
+
+def main(**kwargs):
```

Comment:
I would personally make this two separate files. One for the helper class and another for the script logic. Seem cleaner and makes for smaller files and chunks of code.

Reply:
If this gets a bit more complicated that makes sense. Seems for now it's fine where it is. One benefit is that the cli is with the class it's using. It also follows the other cli tools and there could be a separate PR that consolidates the cli commands from various places.